File: apps/backend/src/modules/emergency-call/domain/entities/emergency-call.entity.ts
import { Location } from '../value-objects/location.vo';
import { CallStatus, CallStatusType } from '../value-objects/call-status.vo';

/**
 * Emergency Call Domain Entity
 * Represents an emergency legal assistance request
 * Follows DDD principles with rich domain logic
 */
export class EmergencyCall {
  private constructor(
    public readonly id: string,
    public readonly userId: string,
    public lawyerId: string | null,
    public readonly location: Location,
    public readonly address: string,
    public status: CallStatus,
    public readonly notes: string | null,
    public readonly createdAt: Date,
    public acceptedAt: Date | null,
    public completedAt: Date | null,
  ) {}

  /**
   * Creates a new emergency call
   */
  static create(
    id: string,
    userId: string,
    location: Location,
    address: string,
    notes: string | null = null,
  ): EmergencyCall {
    return new EmergencyCall(
      id,
      userId,
      null,
      location,
      address,
      CallStatus.pending(),
      notes,
      new Date(),
      null,
      null,
    );
  }

  /**
   * Accepts the emergency call by a lawyer
   */
  accept(lawyerId: string): void {
    if (!this.status.isPending()) {
      throw new Error(
        `Cannot accept call with status: ${this.status.toString()}`,
      );
    }

    if (!lawyerId) {
      throw new Error('Lawyer ID is required to accept call');
    }

    this.lawyerId = lawyerId;
    this.status = CallStatus.accepted();
    this.acceptedAt = new Date();
  }

  /**
   * Completes the emergency call
   */
  complete(): void {
    if (!this.status.isAccepted()) {
      throw new Error(
        `Cannot complete call with status: ${this.status.toString()}`,
      );
    }

    this.status = CallStatus.completed();
    this.completedAt = new Date();
  }

  /**
   * Cancels the emergency call
   */
  cancel(): void {
    if (this.status.isCompleted()) {
      throw new Error('Cannot cancel a completed call');
    }

    if (this.status.isCancelled()) {
      throw new Error('Call is already cancelled');
    }

    this.status = CallStatus.cancelled();
  }

  /**
   * Checks if call can be accepted
   */
  canBeAccepted(): boolean {
    return this.status.isPending();
  }

  /**
   * Checks if call can be completed
   */
  canBeCompleted(): boolean {
    return this.status.isAccepted();
  }

  /**
   * Gets call duration in minutes
   */
  getDurationInMinutes(): number | null {
    if (!this.acceptedAt || !this.completedAt) {
      return null;
    }

    const duration = this.completedAt.getTime() - this.acceptedAt.getTime();
    return Math.floor(duration / (1000 * 60));
  }

  /**
   * Checks if call is active
   */
  isActive(): boolean {
    return this.status.isAccepted() || this.status.isPending();
  }

  /**
   * Converts to plain object
   */
  toObject(): {
    id: string;
    userId: string;
    lawyerId: string | null;
    latitude: number;
    longitude: number;
    address: string;
    status: string;
    notes: string | null;
    createdAt: Date;
    acceptedAt: Date | null;
    completedAt: Date | null;
  } {
    return {
      id: this.id,
      userId: this.userId,
      lawyerId: this.lawyerId,
      latitude: this.location.latitude,
      longitude: this.location.longitude,
      address: this.address,
      status: this.status.toString(),
      notes: this.notes,
      createdAt: this.createdAt,
      acceptedAt: this.acceptedAt,
      completedAt: this.completedAt,
    };
  }
}

File: apps/mobile/lib/features/emergency_call/domain/entities/emergency_call.entity.dart
import 'package:equatable/equatable.dart';

/// Represents an emergency call for immediate legal assistance
/// Domain entity following DDD principles
class EmergencyCallEntity extends Equatable {
  final String id;
  final String userId;
  final String? lawyerId;
  final double latitude;
  final double longitude;
  final String address;
  final EmergencyCallStatus status;
  final String? notes;
  final DateTime createdAt;
  final DateTime? acceptedAt;
  final DateTime? completedAt;

  const EmergencyCallEntity({
    required this.id,
    required this.userId,
    this.lawyerId,
    required this.latitude,
    required this.longitude,
    required this.address,
    required this.status,
    this.notes,
    required this.createdAt,
    this.acceptedAt,
    this.completedAt,
  });

  /// Creates a new emergency call
  factory EmergencyCallEntity.create({
    required String id,
    required String userId,
    required double latitude,
    required double longitude,
    required String address,
    String? notes,
  }) {
    return EmergencyCallEntity(
      id: id,
      userId: userId,
      latitude: latitude,
      longitude: longitude,
      address: address,
      status: EmergencyCallStatus.pending,
      notes: notes,
      createdAt: DateTime.now(),
    );
  }

  /// Accepts the emergency call by a lawyer
  EmergencyCallEntity accept(String lawyerId) {
    return copyWith(
      lawyerId: lawyerId,
      status: EmergencyCallStatus.accepted,
      acceptedAt: DateTime.now(),
    );
  }

  /// Completes the emergency call
  EmergencyCallEntity complete() {
    return copyWith(
      status: EmergencyCallStatus.completed,
      completedAt: DateTime.now(),
    );
  }

  /// Cancels the emergency call
  EmergencyCallEntity cancel() {
    return copyWith(status: EmergencyCallStatus.cancelled);
  }

  /// Creates a copy with updated fields
  EmergencyCallEntity copyWith({
    String? id,
    String? userId,
    String? lawyerId,
    double? latitude,
    double? longitude,
    String? address,
    EmergencyCallStatus? status,
    String? notes,
    DateTime? createdAt,
    DateTime? acceptedAt,
    DateTime? completedAt,
  }) {
    return EmergencyCallEntity(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      lawyerId: lawyerId ?? this.lawyerId,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      address: address ?? this.address,
      status: status ?? this.status,
      notes: notes ?? this.notes,
      createdAt: createdAt ?? this.createdAt,
      acceptedAt: acceptedAt ?? this.acceptedAt,
      completedAt: completedAt ?? this.completedAt,
    );
  }

  @override
  List<Object?> get props => [
        id,
        userId,
        lawyerId,
        latitude,
        longitude,
        address,
        status,
        notes,
        createdAt,
        acceptedAt,
        completedAt,
      ];
}

/// Status of an emergency call
enum EmergencyCallStatus {
  /// Call is pending, waiting for lawyer acceptance
  pending,

  /// Call has been accepted by a lawyer
  accepted,

  /// Call has been completed
  completed,

  /// Call has been cancelled
  cancelled,
}

/// Extension to convert status enum to/from string
extension EmergencyCallStatusExtension on EmergencyCallStatus {
  String toValue() {
    switch (this) {
      case EmergencyCallStatus.pending:
        return 'pending';
      case EmergencyCallStatus.accepted:
        return 'accepted';
      case EmergencyCallStatus.completed:
        return 'completed';
      case EmergencyCallStatus.cancelled:
        return 'cancelled';
    }
  }

  static EmergencyCallStatus fromValue(String value) {
    switch (value) {
      case 'pending':
        return EmergencyCallStatus.pending;
      case 'accepted':
        return EmergencyCallStatus.accepted;
      case 'completed':
        return EmergencyCallStatus.completed;
      case 'cancelled':
        return EmergencyCallStatus.cancelled;
      default:
        throw ArgumentError('Invalid emergency call status: $value');
    }
  }
}
